Add model doc for ViTPose with quantization and attention visualization (#37089)
huggingface/transformers: docs/source/en/model_doc/vitpose.md

person_model = AutoModelForObjectDetection.from_pretrained("PekingU/rtdetr_r50vd_coco_o365").to("cuda")

inputs = person_image_processor(images=image, return_tensors="pt").to("cuda")
with torch.no_grad():
    outputs = person_model(**inputs)

results = person_image_processor.post_process_object_detection(
    outputs, target_sizes=torch.tensor([(image.height, image.width)]), threshold=0.3
)

person_boxes = results[0]["boxes"][results[0]["labels"] == 0].cpu().numpy()
person_boxes[:, 2] -= person_boxes[:, 0]
person_boxes[:, 3] -= person_boxes[:, 1]

# Step 2: Run ViTPose on the cropped person(s)
pose_processor = AutoProcessor.from_pretrained("usyd-community/vitpose-base-simple")
pose_model = AutoModelForPoseEstimation.from_pretrained("usyd-community/vitpose-base-simple").to("cuda")

pose_inputs = pose_processor(image, boxes=[person_boxes], return_tensors="pt").to("cuda")
with torch.no_grad():
    pose_outputs = pose_model(**pose_inputs)

pose_results = pose_processor.post_process_pose_estimation(pose_outputs, boxes=[person_boxes])
```

</hfoption>
</hfoptions>

---

## Quantization Example

```bash
pip install torchao
```

```python

> import torch

Reviewer: You can use the `TorchAoConfig` here instead, see the [Gemma 3](https://huggingface.co/docs/transformers/main/en/model_doc/gemma3#gemma-3) for an example

from torchao.quantization import quantize
from transformers import AutoModelForPoseEstimation

model = AutoModelForPoseEstimation.from_pretrained("usyd-community/vitpose-base-simple")
quantized_model = quantize(model, dtype=torch.qint8)
torch.save(quantized_model.state_dict(), "vitpose-quantized.pt")
```

---

> ## Keypoint Visualization

Reviewer: This should go under the `## Notes` section and it doesn't need its own separate section level

<hfoptions id="visualization">

> <hfoption id="Supervision">

Reviewer: ```suggestion
<hfoption id="supervision">
```

```python
import supervision as sv

annotator = sv.KeypointAnnotator()
image_with_kpts = annotator.annotate(image=np.array(image), keypoints=pose_results[0]["keypoints"])
Image.fromarray(image_with_kpts)
```

</hfoption>
<hfoption id="OpenCV">

```python
import cv2

np_img = np.array(image).copy()
kpts = pose_results[0]["keypoints"]
for x, y, score in kpts:
    if score > 0.3:
        cv2.circle(np_img, (int(x), int(y)), 3, (0, 255, 0), -1)

cv2.imshow("Pose", np_img)
cv2.waitKey(0)
```

</hfoption>
</hfoptions>

---

## Notes

- ViTPose is a top-down pose estimator, meaning it requires a separate object detector to first locate individuals in the image.

> - Use `AutoProcessor` to automatically prepare bounding box and image inputs.

Reviewer: ```suggestion
- Use [`AutoProcessor`] to automatically prepare bounding box and image inputs.
```

> - ViTPose++ supports passing a `dataset_index` parameter to specify which expert head to use.

Reviewer: Recommend clarifying the `dataset_index` parameter and how they relate to the expert heads. See my comment [here](https://github.com/huggingface/transformers/pull/37089#discussion_r2021544579)

> - ViTPose++ supports passing a `dataset_index` parameter to specify which expert head to use.

Reviewer: Add a code snippet here demonstrating using `dataset_index` with ViTPose++ models to fix #36773!

```python
outputs = pose_model(**pose_inputs, dataset_index=2)  # 0: COCO, 1: AiC, 2: MPII, etc.
```

---

## VitPoseImageProcessor

[[autodoc]] VitPoseImageProcessor

>   - preprocess

Reviewer: No need to change the indentation level of these methods

specific language governing permissions and limitations under the License.
-->

<div style="float: right;">
  <div class="flex flex-wrap space-x-1">
    <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">
    <img alt="SDPA" src="https://img.shields.io/badge/SDPA-DE3412?style=flat&logo=pytorch&logoColor=white">
  </div>
</div>

# ViTPose

[ViTPose](https://arxiv.org/abs/2204.12484) is a vision transformer-based model for keypoint (pose) estimation, achieving state-of-the-art performance on MS COCO. It uses a simple, non-hierarchical ViT backbone and lightweight decoder head. Its follow-up, [ViTPose++](https://arxiv.org/abs/2212.04246), incorporates Mixture-of-Experts (MoE) for stronger generalization.

> [!TIP]
> Click on the ViTPose models in the right sidebar for more examples of how to apply ViTPose to pose estimation tasks.

> ---

Reviewer: This hasn't been resolved yet!

## Usage

<hfoptions id="usage">
<hfoption id="Pipeline">

```python
from transformers import pipeline

pose_estimator = pipeline(
    task="pose-estimation",
    model="usyd-community/vitpose-base-simple"

> )

Reviewer 2: We dont have such a pipeline

results = pose_estimator("http://images.cocodataset.org/val2017/000000000139.jpg")
```

</hfoption>
<hfoption id="AutoModel">

```python
import torch
from PIL import Image
from transformers import AutoProcessor, VitPoseForPoseEstimation

image = Image.open("path/to/image.jpg")
processor = AutoProcessor.from_pretrained("usyd-community/vitpose-base-simple")
model = VitPoseForPoseEstimation.from_pretrained("usyd-community/vitpose-base-simple")

inputs = processor(image, return_tensors="pt")
with torch.no_grad():
    outputs = model(**inputs)
```

</hfoption>
<hfoption id="transformers-cli">

```bash
transformers-cli run --task pose-estimation --model usyd-community/vitpose-base-simple --image path/to/image.jpg
```

</hfoption>
</hfoptions>

> ---

Reviewer: ```suggestion
```

Reviewer: Need to remove this too

> ## Quantization Example

Reviewer: Probably better to use a different quantization backend, like torchao, that isn't specific to Intel hardware. Remember to add the complete code snippet for using quantization!

```suggestion
Quantization reduces the memory burden of large models by representing the weights in a lower precision. Refer to the [Quantization](https://huggingface.co/docs/transformers/main/en/quantization/overview) overview for more available quantization backends.

The example below uses [torchao](https://huggingface.co/docs/transformers/main/en/quantization/torchao) to only quantize the weights to int4.
```

Reviewer: Remove `## Quantization Example` and add the text above!

```bash
pip install optimum
```

```python
from optimum.intel import IncQuantizer

quantizer = IncQuantizer.from_pretrained("usyd-community/vitpose-base-simple")
quantizer.quantize(save_directory="./vitpose-quantized")
```

> ---

Reviewer: ```suggestion
```

Reviewer: Remove!

> ## Attention Visualization

Reviewer: The attention mask visualizer isn't supported for ViTPose yet

```python
from transformers.utils.attention_mask_visualizer import AttentionMaskVisualizer

visualizer = AttentionMaskVisualizer(model)
visualizer.visualize(inputs["attention_mask"])
```

> ---

Reviewer: ```suggestion
```

Reviewer: Remove!

## Notes

> - ViTPose is a top-down pose estimator: it requires a prior object detector to crop individuals before keypoint prediction.

Reviewer: ```suggestion
- ViTPose is a top-down pose estimator. It uses a object detector to detect individuals first before keypoint prediction.
```

> - ViTPose++ supports passing a `dataset_index` to activate different MoE expert heads.

Reviewer: Add links to the datasets and include an indented code snippet demonstrating how to pass the `dataset_index`

```suggestion
- ViTPose++ has 6 different MoE expert heads (COCO validation, AiC, MPII, AP-10K, APT-36K, COCO-WholeBody) which supports 6 different datasets. You can pass a specific dataset index to the model to indicate which expert to use.
```

> - Use `AutoProcessor` to handle bounding box preprocessing automatically.
> - Ideal for use cases involving MS COCO, WholeBody, MPII, AP-10K, and APT-36K keypoint datasets.

Reviewer: Use the `<hfoptions>...</hfoptions>` tags under this to add code snippets for supervision and OpenCV.

```suggestion
- Use the [supervision](https://github.com/roboflow/supervision) or [OpenCV](https://opencv.org/) libraries to visualize the keypoints.
```

> ---

Reviewer: ```suggestion
```

Reviewer: Remove!

## VitPoseForPoseEstimation

[[autodoc]] VitPoseForPoseEstimation
  - forward

> ---

Reviewer: Remove and the citation under it!
